SingularityService/src/main/java/com/hubspot/singularity/resources/TaskResource.java

```diff
@@ -180,6 +183,19 @@ public SingularityTaskRequest getPendingTask(@Auth SingularityUser user, @PathPa
     return taskRequestList.get(0);
   }
 
+  @DELETE
+  @Path("/scheduled/task/{scheduledTaskId}")
+  @ApiOperation("Delete a scheduled task.")
+  public void deleteScheduledTask(@Auth SingularityUser user, @PathParam("scheduledTaskId") String taskId) {
+    SingularityTaskRequest taskRequest = getPendingTask(user, taskId);
+    SingularityRequest request = taskRequest.getRequest();
+
+    authorizationHelper.checkForAuthorization(request, user, SingularityAuthorizationScope.ADMIN);
+    checkBadRequest(request.getRequestType() == RequestType.ON_DEMAND, "Only ON_DEMAND tasks may be deleted.");
+
+    taskManager.deletePendingTask(taskRequest.getPendingTask().getPendingTaskId());
```

Comment:
This is a very basic way of doing this. However, it opens us up to some odd cases of inconsistent state with the pollers. Here's an example:

- Two Singularity scheduler instances are running, `1` is the leader `2` is just serving API calls
- The leader grabs the list of pending tasks from zk to schedule them
- API call comes in to delete pending task `123` on instance `2`, and is removed from zk
- The leader may schedule it, or maybe fail to, and if it tries to delete it is no longer there

For this _particular_ case, I think we would still get the desired outcome in the end. However, it is good to think about how readable and debuggable this is. An alternative solution is to create a new node in zk using the TaskManager, for pending tasks that are marked as deleted. Then, on the next run of the poller (which normally consumes pending tasks), the actual delete can take place. This keeps all of the change of state with the leader to make it more consistent and easier to reason about if something were to go wrong and we had to debug it.

Reply:
To clarify, is the code here correctly limiting state change to the leader? i.e. Is debugability/semantic cohesiveness when looking at the zk structure the primary concern?

Reply:
In general Singularity's model of the pollers is based on the fact that certain things can only be edited/deleted by the leading instance. This way the state is easier to manage since we don't have to think about another server editing something we just loaded into memory to operate on.

The code here as it stands will run on whatever instance gets hit by the api call, which may or may not be the current leader. Even if we hit the leader we could have problems of state in that the api call would run on a different thread than the poller. So, the api call thread could be deleting while the poller is operating on the same data.

So the primary concern in this case is keeping the state (i.e. list of pending/active tasks) clean by only allowing the leader to operate it in a locked way.